### PR TITLE
Lunge modification

### PR DESCRIPTION
**Pull request by Undid-Iridium**
At least allows damage refusal and other things. @louis1706 @Thundermaker300 thoughts

### CODE REVIEW

File: Exiled.Events/Handlers/Scp939.cs [hunk 1]
```diff
@@ -25,7 +25,7 @@ public static class Scp939
         /// <summary>
         ///     Invoked before SCP-939 uses its lunge ability.
```
Comment (louis1706) [inline Exiled.Events/Handlers/Scp939.cs:26]: After

File: Exiled.Events/Handlers/Scp939.cs [hunk 2]
```diff
@@ -56,8 +56,8 @@ public static class Scp939
         /// <summary>
         ///     Called before SCP-939 uses its lunge ability.
```
Comment (louis1706) [inline Exiled.Events/Handlers/Scp939.cs:57]: After

**Review by louis1706 — CHANGES_REQUESTED** [review]

File: Exiled.Events/EventArgs/Scp939/LungedEventArgs.cs [hunk 3]
```diff
@@ -9,35 +9,45 @@ namespace Exiled.Events.EventArgs.Scp939
...
-        public bool IsAllowed { get; set; }
+        public Player Target { get; }
 
         /// <summary>
         ///     Gets the player who's controlling SCP-939.
         /// </summary>
         public Player Player { get; }
+
+        /// <summary>
+        /// Gets or sets scp939's lunge state. If you deny event, lunge state changes to this.
```
Comment (louis1706) [inline Exiled.Events/EventArgs/Scp939/LungedEventArgs.cs:49]: you can't deny this event ? now

**Review by louis1706 — CHANGES_REQUESTED** [review]

File: Exiled.Events/Patches/Events/Scp939/Lunge.cs [hunk 4]
```diff
@@ -7,92 +7,127 @@
...
     ///     Patches <see cref="Scp939LungeAbility.ServerProcessCmd(NetworkReader)" />
     ///     to add the <see cref="Scp939" /> event.
     /// </summary>
-    [HarmonyPatch(typeof(Scp939LungeAbility), nameof(Scp939LungeAbility.ServerProcessCmd))]
+    [HarmonyPatch]
     internal static class Lunge
     {
-        private static bool Prefix(Scp939LungeAbility __instance, NetworkReader reader)
+        [HarmonyPatch(typeof(Scp939LungeAbility), nameof(Scp939LungeAbility.ServerProcessCmd))]
+        [HarmonyTranspiler]
```
Comment (iopietro) [inline Exiled.Events/Patches/Events/Scp939/Lunge.cs:38]: This should be done like other patches.

**Review by iopietro — CHANGES_REQUESTED** [review]

File: Exiled.Events/Patches/Events/Scp939/Lunge.cs [hunk 5]
```diff
@@ -26,73 +34,98 @@ namespace Exiled.Events.Patches.Events.Scp939
...
-                    return false;
-
-                LungingEventArgs ev = new(__instance.Owner, __instance.IsReady);
-                Handlers.Scp939.OnLunging(ev);
-                if (!ev.IsAllowed)
-                    return false;
+                if (!instance.IsReady)
+                {
+                    return;
+                }
```
Comment (iopietro) [inline Exiled.Events/Patches/Events/Scp939/Lunge.cs:74]:
```suggestion
                if (!instance.IsReady)
                    return;
```

File: Exiled.Events/Patches/Events/Scp939/Lunge.cs [hunk 6]
```diff
@@ -26,73 +34,98 @@ namespace Exiled.Events.Patches.Events.Scp939
...
+            ListPool<CodeInstruction>.Pool.Return(newInstructions);
+        }
+
+        private static void ProcessLunge(Scp939LungeAbility instance, NetworkReader reader)
         {
             Vector3 vector = reader.ReadRelativePosition().Position;
             ReferenceHub referenceHub = reader.ReadReferenceHub();
             RelativePosition relativePosition = reader.ReadRelativePosition();
-            if (__instance.State != Scp939LungeState.Triggered)
+            if (instance.State != Scp939LungeState.Triggered)
```
Comment (iopietro) [inline Exiled.Events/Patches/Events/Scp939/Lunge.cs:69]:
```suggestion

            if (instance.State != Scp939LungeState.Triggered)
```

File: Exiled.Events/Patches/Events/Scp939/Lunge.cs [hunk 7]
```diff
@@ -26,73 +34,98 @@ namespace Exiled.Events.Patches.Events.Scp939
...
-                __instance.TriggerLunge();
+                instance.TriggerLunge();
             }
 
             HumanRole humanRole;
             if (referenceHub == null || (humanRole = referenceHub.roleManager.CurrentRole as HumanRole) == null)
             {
-                return false;
+                return;
             }
```
Comment (iopietro) [inline Exiled.Events/Patches/Events/Scp939/Lunge.cs:83]:
```suggestion
```

File: Exiled.Events/Patches/Events/Scp939/Lunge.cs [hunk 8]
```diff
@@ -26,73 +34,98 @@ namespace Exiled.Events.Patches.Events.Scp939
...
             Vector3 position = fpcModule.Position;
             Quaternion rotation = transform.rotation;
             Vector3 vector3 = new(vector.x, position.y, vector.z);
-            transform.forward = -__instance.Owner.transform.forward;
+            transform.forward = -instance.Owner.transform.forward;
             fpcModule.Position = vector3;
-            bool flag = referenceHub.playerStats.DealDamage(new Scp939DamageHandler(__instance.ScpRole, Scp939DamageType.LungeTarget));
+            bool flag = referenceHub.playerStats.DealDamage(new Scp939DamageHandler(instance.ScpRole, Scp939DamageType.LungeTarget));
             float num = flag ? 1f : 0f;
             if (!flag || referenceHub.IsAlive())
```
Comment (iopietro) [inline Exiled.Events/Patches/Events/Scp939/Lunge.cs:116]:
```suggestion

            if (!flag || referenceHub.IsAlive())
```

File: Exiled.Events/Patches/Events/Scp939/Lunge.cs [hunk 9]
```diff
@@ -101,11 +134,12 @@ private static bool Prefix(Scp939LungeAbility __instance, NetworkReader reader)
 
             if (flag)
             {
-                Hitmarker.SendHitmarker(__instance.Owner, num);
+                Hitmarker.SendHitmarker(instance.Owner, num);
             }
```
Comment (iopietro) [inline Exiled.Events/Patches/Events/Scp939/Lunge.cs:138]:
```suggestion
```

File: Exiled.Events/Patches/Events/Scp939/Lunge.cs [hunk 10]
```diff
@@ -26,73 +34,98 @@ namespace Exiled.Events.Patches.Events.Scp939
...
-        private static bool Prefix(Scp939LungeAbility __instance, NetworkReader reader)
+        private static IEnumerable<CodeInstruction> OnLungeHit(IEnumerable<CodeInstruction> instructions, ILGenerator generator)
+        {
+            List<CodeInstruction> newInstructions = ListPool<CodeInstruction>.Pool.Get(instructions);
+
+            Label returnLabel = generator.DefineLabel();
+
+            // Right after base call.
+            int index = newInstructions.FindIndex(instruction => instruction.Calls(Method(typeof(ScpSubroutineBase), nameof(ScpSubroutineBase.ServerProcessCmd)))) + 1;
+            newInstructions.InsertRange(
```
Comment (iopietro) [inline Exiled.Events/Patches/Events/Scp939/Lunge.cs:45]:
```suggestion
            int index = newInstructions.FindIndex(instruction => instruction.Calls(Method(typeof(ScpSubroutineBase), nameof(ScpSubroutineBase.ServerProcessCmd)))) + 1;

            newInstructions.InsertRange(
```

**Review by iopietro — CHANGES_REQUESTED** [review]

File: Exiled.Events/Patches/Events/Scp939/Lunge.cs [hunk 11]
```diff
@@ -26,86 +34,108 @@ namespace Exiled.Events.Patches.Events.Scp939
...
+
+            newInstructions.InsertRange(
+                index,
+                new[]
+                {
+                    // Scp939ProcessLunge, NetworkReader
+                    new CodeInstruction(OpCodes.Ldarg_0),
+                    new CodeInstruction(OpCodes.Ldarg_1),
+                    new(OpCodes.Call, Method(typeof(Lunge), nameof(ProcessLunge))),
+                    new(OpCodes.Br, returnLabel),
```
Comment (iopietro) [inline Exiled.Events/Patches/Events/Scp939/Lunge.cs:54]:
```suggestion
                    new(OpCodes.Br_S, returnLabel),
```

**Review by iopietro — CHANGES_REQUESTED** [review]

**Review by iopietro — APPROVED** [review]

**Review by louis1706 — CHANGES_REQUESTED** [review]
Make LungingEventArgs / OnLunging Obsolete instead of renaming them

Obsolete["Replace by LungingEventArgs", true]

**Comment by Undid-Iridium** [comment]
> Make LungingEventArgs / OnLunging Obsolete instead of renaming them
> 
> Obsolete["Replace by LungingEventArgs", true]

image

Make the change you want and I'll commit it like Pietro does

**Comment by louis1706** [comment]
Also the event will need a rework to be a real transpiler

**Comment by Undid-Iridium** [comment]
Since I can't seem to tag it, Needs testing.

**Comment by Undid-Iridium** [comment]
Someone else can take this

**Pull request closed without merging**